src/gui/editors/PianoRoll.cpp

```diff
@@ -662,6 +663,73 @@ void PianoRoll::clearGhostPattern()
 }
 
 
+void PianoRoll::glueNotes()
+{
+	if (hasValidPattern())
+	{
+		NoteVector selectedNotes = getSelectedNotes();
+		if (selectedNotes.empty())
+		{
+			TextFloat::displayMessage( tr( "Glue notes failed" ),
+					tr( "Please select notes to glue first." ),
+					embed::getIconPixmap( "glue", 24, 24 ),
+					3000 );
+			return;
+		}
+
+		// Make undo possible
+		m_pattern->addJournalCheckPoint();
+
+		// Sort notes on key and then pos.
+		std::sort(selectedNotes.begin(), selectedNotes.end(),
+			[](const Note * note, const Note * compareNote) -> bool
+			{
+				if (note->key() == compareNote->key())
+				{
+					return note->pos() < compareNote->pos();
+				}
+				return note->key() < compareNote->key();
+			});
+
+		QList<Note *> noteToRemove;
+
+		NoteVector::iterator note = selectedNotes.begin();
+		auto nextNote = note+1;
+		NoteVector::iterator end = selectedNotes.end();
+
+		while (note != end && nextNote != end)
+		{
+			// key and position match for glue
+			if ((*note)->key() == (*nextNote)->key()
+				&& (*nextNote)->pos() >= (*note)->pos()
+				&& (*nextNote)->pos() <= (*note)->pos()
+				+ (*note)->length())
```

Comment:
`(*note)->length()` Maybe change to something like: `(*note)->length() < 0 ? 0 : (*note)->length()`

Reply:
Maybe add a `(*note)->length() > 0` in the if statement to reject beat notes entirely from glue? Otherwise you'd also need similar guard logic when getting the end of the next note. Also, setting beat note length to zero for this calculation could look weird since they're drawn as four ticks long. Might lead to two notes not being merged although they look connected.

Reply:
Yes, that sounds better.

Reply:
In FL studio, normal notes absorb infinite length notes(similar to our BB notes).